Implement miniogw in terms of ObjectStore

aligeti opened the pull request:
Implement miniogw in terms of ObjectStore

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/storj/storj/117)
<!-- Reviewable:end -->

kaloyan-raev commented:
Review status: 0 of 4 files reviewed, 13 unresolved discussions (waiting on @aligeti and @jtolds)

---

*[pkg/miniogw/gateway-storj.go, line 44 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjLFH9LR04lUu1kjl:-LGQjLFH9LR04lUu1kjm:b-1igrjh) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L44)):*
> ```Go
> 
> // getBuckets returns the buckets list
> func (s *storjObjects) getBuckets() (buckets []minio.BucketInfo, err error) {
> ```

It's time to delete this function.

---

*[pkg/miniogw/gateway-storj.go, line 56 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjNqkCTvOP1zhKDe5:-LGQjNqkCTvOP1zhKDe6:b-1igrjh) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L56)):*
> ```Go
> 
> // uploadFile function handles to add the uploaded file to the bucket's file list structure
> func (s *storjObjects) uploadFile(bucket, object string, filesize int64, metadata map[string]string) (result minio.ListObjectsInfo, err error) {
> ```

It's time to delete this function.

---

*[pkg/miniogw/gateway-storj.go, line 97 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjR2k7jFY_lCi3JOc:-LGQjR2k7jFY_lCi3JOd:b-1igrjh) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L97)):*
> ```Go
> 
> // getFiles returns the files list for a bucket
> func (s *storjObjects) getFiles(bucket string) (result minio.ListObjectsInfo, err error) {
> ```

It's time to delete this function.

---

*[pkg/miniogw/gateway-storj.go, line 147 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQk03V3Nta9DthKq4m:-LGQk03X70xes2EvKt2a:b42dluk) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L147)):*
> ```Go
> // Storj is the implementation of a minio cmd.Gateway
> type Storj struct {
> 	bucketlist []S3Bucket
> ```

This bucket list should be removed. The gateway itself should not store any bucket/object info. This should be done only via the `ObjectStore` interface.

---

*[pkg/miniogw/gateway-storj.go, line 168 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQkLsyCg-sAgeSzrVb:-LGQkLsyCg-sAgeSzrVc:b-1igrjh) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L168)):*
> ```Go
> 
> //createSampleBucketList function initializes sample buckets and files in each bucket
> func (s *Storj) createSampleBucketList() {
> ```

It's time to delete this function.

---

*[pkg/miniogw/gateway-storj.go, line 197 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQZsAX3hunp3Dgvaqc:-LGQZsAX3hunp3Dgvaqd:b-2fv2hm) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L197)):*
> ```Go
> func (s *storjObjects) DeleteObject(ctx context.Context, bucket,
> 	object string) error {
> 	return s.storj.os.DeleteObject(ctx, object)
> ```

The path passed to the object store should be a concatenation of the `bucket` and `object`.

---

*[pkg/miniogw/gateway-storj.go, line 207 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQ_1ni7HGoewxNQCf5:-LGQ_1nk-Vyymvfj5Ox6:b-3mdtn9) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L207)):*
> ```Go
> func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
> 	startOffset int64, length int64, writer io.Writer, etag string) (err error) {
> 	_, _, err = s.storj.os.GetObject(ctx, object)
> ```

The path passed to the object store should be a concatenation of the `bucket` and `object`.

You also need to copy the data to the `writer`. Otherwise the client does not receive anything.

---

*[pkg/miniogw/gateway-storj.go, line 214 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQhn4oDNqvF9mR5yev:-LGQhn4oDNqvF9mR5yew:b-i6ab3b) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L214)):*
> ```Go
> 	object string) (objInfo minio.ObjectInfo, err error) {
> 	_, m, err := s.storj.os.GetObject(ctx, object)
> 	objInfo = minio.ObjectInfo{
> ```

Only bucket in the ObjectInfo? In the demo branch we had quite more info: https://github.com/storj/storj/blob/demo/pkg/miniogw/gateway-storj.go#L301-L310

---

*[pkg/miniogw/gateway-storj.go, line 228 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjZdo1atxBYTF4ING:-LGQjZdo1atxBYTF4INH:br8nbr6) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L228)):*
> ```Go
> 	delimiter string, maxKeys int) (result minio.ListObjectsInfo, err error) {
> 	_, _, _ = s.storj.os.ListObjects(ctx, prefix, delimiter)
> 	return s.getFiles(bucket)
> ```

This result should be taken from `ObjectStore.ListObjects` and not from the fake `getFiles`.

---

*[pkg/miniogw/gateway-storj.go, line 237 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjATa-6EnARZdh8YK:-LGQjATa-6EnARZdh8YL:b-1igrjh) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L237)):*
> ```Go
> 
> //encryptFile encrypts the uploaded files
> func encryptFile(data io.ReadCloser, blockSize uint, bucket, object string) error {
> ```

It's time to delete this function.

---

*[pkg/miniogw/gateway-storj.go, line 289 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQitkDC4zHYj413UWN:-LGQitkDC4zHYj413UWO:bqr55ww) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L289)):*
> ```Go
> 	err error) {
> 	objPath := filepath.Join(bucket, object)
> 	wsize, err := s.storj.os.PutObject(ctx, objPath, data, nil, time.Now())
> ```

How did it happen that `ObjectStore.PutObject` not returns `size int64`?

Also you are not passing the `metadata` to `ObjectStore.PutObject`.

---

*[pkg/miniogw/gateway-storj.go, line 291 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQlNNPCUWHwCWEGqaL:-LGQlNNPCUWHwCWEGqaM:by4azzv) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L291)):*
> ```Go
> 	wsize, err := s.storj.os.PutObject(ctx, objPath, data, nil, time.Now())
> 	if err == nil {
> 		s.uploadFile(bucket, object, wsize, metadata)
> ```

This old function should not be called anymore.

---

*[pkg/objects/iface.go, line 25 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQi5-H9T4eeGHZ5VTt:-LGQi5-H9T4eeGHZ5VTu:bi5mvnb) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/objects/iface.go#L25)):*
> ```Go
> //ObjectStore interface
> type ObjectStore interface {
> 	PutObject(ctx context.Context, path string, data io.Reader, metadata []byte, expiration time.Time) (size int64, err error)
> ```

The Path type was merged. You should now use it for paths instead of `string`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

aligeti commented:
*[pkg/miniogw/gateway-storj.go, line 228 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjZdo1atxBYTF4ING:-LGQzTFC1ERq0OR658iR:b-r23ogh) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L228)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

This result should be taken from `ObjectStore.ListObjects` and not from the fake `getFiles`.
</blockquote></details>

I am aware of that.... I left it there just to keep the minio to behave like as it was before. Not sure should I change the behavior of minio client for now? its a question to @jt.  My recommedation is to hold on to any changes to miniogw until all the pieces are lined up. Well lets ask @JT.

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

aligeti commented:
Review status: 0 of 4 files reviewed, 13 unresolved discussions (waiting on @aligeti, @kaloyan-raev, and @jtolds)

---

*[pkg/miniogw/gateway-storj.go, line 44 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjLFH9LR04lUu1kjl:-LGQrb864ZxUefdYRi6K:b-2kkzow) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L44)):*
> The path passed to the object store should be a concatenation of the bucket and object.

---

*[pkg/miniogw/gateway-storj.go, line 56 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjNqkCTvOP1zhKDe5:-LGQroXa5Fv00XPlYNcU:b-8yai3i) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L56)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

It's time to delete this function.
</blockquote></details>

 I know there are lot of hardcoded functions which fills in the buckets and files (objects). Should be start cleaning up the miniogw code?

---

*[pkg/miniogw/gateway-storj.go, line 97 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjR2k7jFY_lCi3JOc:-LGQsBxGDRLL_Nj5qFgn:b-gj1q4e) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L97)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

It's time to delete this function.
</blockquote></details>

 I know there are lot of hardcoded functions which fills in the buckets and files (objects). Should  start cleaning up the miniogw code, now or should we wait until the streamobjecct and segment object are in place?

---

*[pkg/miniogw/gateway-storj.go, line 147 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQk03V3Nta9DthKq4m:-LGQsHFP3sFvxUrU9DQV:b9tva0o) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L147)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

This bucket list should be removed. The gateway itself should not store any bucket/object info. This should be done only via the `ObjectStore` interface.
</blockquote></details>

till the time being should we send the hardcoded data of buckets and objects from objectstore? basically move the getbuckets(), get files() functionality to Objectstore?

---

*[pkg/miniogw/gateway-storj.go, line 168 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQkLsyCg-sAgeSzrVb:-LGQsXL1ASCEz6RjTApF:b1jet0p) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L168)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

It's time to delete this function.
</blockquote></details>

same comment as above.

---

*[pkg/miniogw/gateway-storj.go, line 197 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQZsAX3hunp3Dgvaqc:-LGQs_7xA43Dia9uMUF8:b-fq04gg) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L197)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

The path passed to the object store should be a concatenation of the `bucket` and `object`.
</blockquote></details>

if my bucket name is "bucket" and object name is" object" concatenation is "bucket object" or its file path.Join() ? how do you want the objectstore to pass data to streamstore? I thought it was file path.JOin() per the JTs writeup. Which one is it?

---

*[pkg/miniogw/gateway-storj.go, line 207 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQ_1ni7HGoewxNQCf5:-LGQxFmyDhPPdNkTNV6q:b-p8y2rk) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L207)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

The path passed to the object store should be a concatenation of the `bucket` and `object`.

You also need to copy the data to the `writer`. Otherwise the client does not receive anything.
</blockquote></details>

func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
	startOffset int64, length int64, writer io.Writer, etag string) (err error) {
	data, _, err = s.storj.os.GetObject(ctx, object)
      	wsize, err := io.CopyN(writer, data, data.Size())
/* here just err either err !=nil or err == nil as the function returns nil */
	return err
}

that is what I will modify. Does that addresses your comment?

---

*[pkg/miniogw/gateway-storj.go, line 214 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQhn4oDNqvF9mR5yev:-LGQyUoB-rokE3Xl2Acl:bu3x12n) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L214)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

Only bucket in the ObjectInfo? In the demo branch we had quite more info: https://github.com/storj/storj/blob/demo/pkg/miniogw/gateway-storj.go#L301-L310
</blockquote></details>

I am aware of the objectInfo, from the demo.  The Object info has to be filled based on what is returned in m, which should be a superset of data. ie It depend on "m", for now I just added some dummy member. I don't know exactly what is returned from the lower layers to objectStore's Meta Struture. I just had the place holder until all pieces comes together. I will for now hard code the objectstore in the GetObjectInfo() to address this comment. 

NOTE: I added some data members to meta structure based on S3 implementation. 

---

*[pkg/miniogw/gateway-storj.go, line 289 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQitkDC4zHYj413UWN:-LGR0C5z6WksmZzqWD03:bifuprn) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L289)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

How did it happen that `ObjectStore.PutObject` not returns `size int64`?

Also you are not passing the `metadata` to `ObjectStore.PutObject`.
</blockquote></details>

I added the ObjectStore.PutObject to return the size. It wasn't there in the @JT spec. Also the  metadata map[string]string how is this data used or passed thru ObjectStore.PutObject() ?  Also minio client passes "content-type" key and empty value  for the arg metadata map[string]string. how does this arg dictates the handling in ObjectStore.PutObject() ?  

---

*[pkg/miniogw/gateway-storj.go, line 291 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQlNNPCUWHwCWEGqaL:-LGR1Z1X-Z4V2JAqt4X0:b-icilga) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L291)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

This old function should not be called anymore.
</blockquote></details>

Same comment. I understand. But should be make the change and remove all the hard coded func demo version function and make miniogw look like an empty page? ofcourse we can emulate the same hardcoded values to be returned from Objectstore. What do you think we should do? should we hard code in objectstore or wait little longer till all the stream/segment store pieces in place? let me know what you think. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

aligeti commented:
Review status: 0 of 4 files reviewed, 13 unresolved discussions (waiting on @kaloyan-raev, @aligeti, and @jtolds)

---

*[pkg/miniogw/gateway-storj.go, line 237 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjATa-6EnARZdh8YK:-LGR2G2V3CzdxowVPdOh:b-tz5zp3) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L237)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

It's time to delete this function.
</blockquote></details>

its technically no use now.. I agree.. will remove that function

---

*[pkg/objects/iface.go, line 25 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQi5-H9T4eeGHZ5VTt:-LGR26c-BPpPpTujVw9e:b-qlfp1d) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/objects/iface.go#L25)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

The Path type was merged. You should now use it for paths instead of `string`.
</blockquote></details>

ok will make the change.. I was waiting for your changes. will make the modes

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

aligeti commented:
Review status: 0 of 4 files reviewed, 13 unresolved discussions (waiting on @kaloyan-raev, @aligeti, and @jtolds)

---

*[pkg/miniogw/gateway-storj.go, line 207 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQ_1ni7HGoewxNQCf5:-LGR2udV6VTq3yqOfl99:b-qf0bib) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L207)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
	startOffset int64, length int64, writer io.Writer, etag string) (err error) {
	data, _, err = s.storj.os.GetObject(ctx, object)
      	wsize, err := io.CopyN(writer, data, data.Size())
/* here just err either err !=nil or err == nil as the function returns nil */
	return err
}

that is what I will modify. Does that addresses your comment?
</blockquote></details>

correction:  return err straight away as the function returns err

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

aligeti commented:
what I am looking at?

---

Review status: 0 of 4 files reviewed, 13 unresolved discussions (waiting on @kaloyan-raev and @jtolds)

---



*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117#-:-LGR3N-JE3jq1dhohNV9:bcyhsvd)*
<!-- Sent from Reviewable.io -->

aligeti commented:
Review status: 0 of 4 files reviewed, 13 unresolved discussions (waiting on @kaloyan-raev and @jtolds)

---

*[pkg/objects/iface.go, line 25 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQi5-H9T4eeGHZ5VTt:-LGRMOws4v9CU9qmvFw_:b-wuuk9a) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/objects/iface.go#L25)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

ok will make the change.. I was waiting for your changes. will make the modes
</blockquote></details>

The ObjectStore's methods are the API to the outside to use and develop proprietary applications using C/C++ , mobile apps and so on. Keeping the arguments to basic datatypes would help in easy integration. I looked at the type Path set to []string, which I would recommend not to use that Path type. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

kaloyan-raev commented:
Review status: 0 of 4 files reviewed, 13 unresolved discussions (waiting on @kaloyan-raev and @jtolds)

---

*[pkg/miniogw/gateway-storj.go, line 197 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQZsAX3hunp3Dgvaqc:-LGUL0a-9KqXGBXVE2_9:b-iayy41) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L197)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

if my bucket name is "bucket" and object name is" object" concatenation is "bucket object" or its file path.Join() ? how do you want the objectstore to pass data to streamstore? I thought it was file path.JOin() per the JTs writeup. Which one is it?
</blockquote></details>

You can simply use `paths.New(bucket, object)` and it will create the correct path to pass to the ObjectStore methods.

---

*[pkg/miniogw/gateway-storj.go, line 207 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQ_1ni7HGoewxNQCf5:-LGUMTyTC1f5XORJiQrF:bmya2bg) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L207)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

correction:  return err straight away as the function returns err
</blockquote></details>

This isin the right direction, although you need to polish the details. `ObjectStore.GetObject` returns a `RangeCloser`. You should get a `ReadCloser` from it and copy it to the gateway's writer. And make sure to close the ReadCloser and RangeCloser.

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

aligeti commented:
Review status: 0 of 4 files reviewed, 12 unresolved discussions (waiting on @aligeti, @kaloyan-raev, and @jtolds)

---

*[pkg/miniogw/gateway-storj.go, line 56 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjNqkCTvOP1zhKDe5:-LGVVKJjEvfIZ-gf7bP-:b50zaeb) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L56)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

 I know there are lot of hardcoded functions which fills in the buckets and files (objects). Should be start cleaning up the miniogw code?
</blockquote></details>

resolution agreed to remove the functions that create dummy data. 

---

*[pkg/miniogw/gateway-storj.go, line 97 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjR2k7jFY_lCi3JOc:-LGVVQ_54WS051CtVbww:b50zaeb) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L97)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

 I know there are lot of hardcoded functions which fills in the buckets and files (objects). Should  start cleaning up the miniogw code, now or should we wait until the streamobjecct and segment object are in place?
</blockquote></details>

resolution agreed to remove the functions that create dummy data. 

---

*[pkg/miniogw/gateway-storj.go, line 147 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQk03V3Nta9DthKq4m:-LGVVTvXDQjyUq8U9Gdp:b50zaeb) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L147)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

till the time being should we send the hardcoded data of buckets and objects from objectstore? basically move the getbuckets(), get files() functionality to Objectstore?
</blockquote></details>

resolution agreed to remove the functions that create dummy data. 

---

*[pkg/miniogw/gateway-storj.go, line 168 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQkLsyCg-sAgeSzrVb:-LGVVVdNBBj29u2zH9ty:b50zaeb) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L168)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

same comment as above.
</blockquote></details>

resolution agreed to remove the functions that create dummy data. 

---

*[pkg/miniogw/gateway-storj.go, line 197 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQZsAX3hunp3Dgvaqc:-LGVWRvCCuA7sIC_H3EL:b-c9b9mp) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L197)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

You can simply use `paths.New(bucket, object)` and it will create the correct path to pass to the ObjectStore methods.
</blockquote></details>

Resolution agreed is don't change the argument type in the interface. Go back to what it was defined in the JT interface spec. So change will be to change from string back to path in the package. Move the files to /pkg/storage/

---

*[pkg/miniogw/gateway-storj.go, line 207 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQ_1ni7HGoewxNQCf5:-LGVaNKt5CRRFvSCRngM:b-12xcn8) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L207)):*
<details><summary><i>Previously, kaloyan-raev (Kaloyan Raev) wrote…</i></summary><blockquote>

This isin the right direction, although you need to polish the details. `ObjectStore.GetObject` returns a `RangeCloser`. You should get a `ReadCloser` from it and copy it to the gateway's writer. And make sure to close the ReadCloser and RangeCloser.
</blockquote></details>

made the changes using the ranger closer and modified. 

---

*[pkg/miniogw/gateway-storj.go, line 214 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQhn4oDNqvF9mR5yev:-LGVb0Tu2rdGDjI-iRko:by7t74f) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L214)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

I am aware of the objectInfo, from the demo.  The Object info has to be filled based on what is returned in m, which should be a superset of data. ie It depend on "m", for now I just added some dummy member. I don't know exactly what is returned from the lower layers to objectStore's Meta Struture. I just had the place holder until all pieces comes together. I will for now hard code the objectstore in the GetObjectInfo() to address this comment. 

NOTE: I added some data members to meta structure based on S3 implementation. 
</blockquote></details>

Also propose protobuf for the meta's []byte. Need to talk to Brandon/JT about implementing the protobuf for meta's []byte

---

*[pkg/miniogw/gateway-storj.go, line 228 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjZdo1atxBYTF4ING:-LGVbwCAD2QcaDaqAvV0:b-4k3v2x) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L228)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

I am aware of that.... I left it there just to keep the minio to behave like as it was before. Not sure should I change the behavior of minio client for now? its a question to @jt.  My recommedation is to hold on to any changes to miniogw until all the pieces are lined up. Well lets ask @JT.
</blockquote></details>

Out of scope of this ticket, just stub it out but add copyObject implement

---

*[pkg/miniogw/gateway-storj.go, line 237 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjATa-6EnARZdh8YK:-LGVcLDe0HD6Nk-5To3V:b50zaeb) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L237)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

its technically no use now.. I agree.. will remove that function
</blockquote></details>

resolution agreed to remove the functions that create dummy data. 

---

*[pkg/miniogw/gateway-storj.go, line 289 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQitkDC4zHYj413UWN:-LGVesTY2DJHiNXrP8Tf:bb4mr6o) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L289)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

I added the ObjectStore.PutObject to return the size. It wasn't there in the @JT spec. Also the  metadata map[string]string how is this data used or passed thru ObjectStore.PutObject() ?  Also minio client passes "content-type" key and empty value  for the arg metadata map[string]string. how does this arg dictates the handling in ObjectStore.PutObject() ?  
</blockquote></details>

need more discussion.

---

*[pkg/miniogw/gateway-storj.go, line 291 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQlNNPCUWHwCWEGqaL:-LGVeuy3-rw2Gy-T1Mws:b50zaeb) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L291)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

Same comment. I understand. But should be make the change and remove all the hard coded func demo version function and make miniogw look like an empty page? ofcourse we can emulate the same hardcoded values to be returned from Objectstore. What do you think we should do? should we hard code in objectstore or wait little longer till all the stream/segment store pieces in place? let me know what you think. 
</blockquote></details>

resolution agreed to remove the functions that create dummy data. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

aligeti commented:
Review status: 0 of 4 files reviewed, 12 unresolved discussions (waiting on @aligeti, @kaloyan-raev, and @jtolds)

---

*[pkg/miniogw/gateway-storj.go, line 197 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQZsAX3hunp3Dgvaqc:-LGVrd81Dd5_5iva8ea0:bu9y57e) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L197)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

Resolution agreed is don't change the argument type in the interface. Go back to what it was defined in the JT interface spec. So change will be to change from string back to path in the package. Move the files to /pkg/storage/
</blockquote></details>

Changed back all the interfaces back to per specification expect ListObjects(ctx context.Context, startingPath, endingPath paths.Path) (paths paths.Path, truncated bool, err error)

JT specified as 	ListObjects(ctx context.Context, startingPath, endingPath dtypes.Path) (paths []dtypes.Path, truncated bool, err error) 

the paths is of type slice, I guess it should have been a typo

Will be moved once ready all code reviews changes are accepted. One thing at a time, first code acceptability and then files will be renamed and moved

---

*[pkg/miniogw/gateway-storj.go, line 207 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQ_1ni7HGoewxNQCf5:-LGVsccg1zg-SAfxS0fY:b-896fix) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L207)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

made the changes using the ranger closer and modified. 
</blockquote></details>

Done.

---

*[pkg/miniogw/gateway-storj.go, line 237 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQjATa-6EnARZdh8YK:-LGVsh0G8U3LnmyTx6lM:b-896fix) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L237)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

resolution agreed to remove the functions that create dummy data. 
</blockquote></details>

Done.

---

*[pkg/miniogw/gateway-storj.go, line 289 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQitkDC4zHYj413UWN:-LGVspQr6itfej2EI0Zn:brbtrn1) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L289)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

need more discussion.
</blockquote></details>

Modified the function declaration as per the JT Spec. The function definition looks as below 

1. func (s *storjObjects) PutObject(ctx context.Context, bucket, object string,
2. 	data *hash.Reader, metadata map[string]string) (objInfo minio.ObjectInfo,
3. 	err error) {
4.	objpath := paths.New(bucket, object)
5.	err = s.storj.os.PutObject(ctx, objpath, data, nil, time.Now())
6.	return minio.ObjectInfo{
7.		Name:    object,
8.		Bucket:  bucket,
9.		ModTime: time.Now(),
10.		Size:    data.Size(),
11.		ETag:    minio.GenETag(),
12. 	}, err
13. }


line 5 is the call made into ObjectStore's PutObject method and it just returns err and when it comes of the function at line5 and goes to line6, we need to fill the Objectinfo to be relayed back to minio. Now the question is how to I fill the Size in line10? If you are suggesting me to take the size of data from line2, then I have a question how do you guarantee that the size of data from line2 is same as from line 5 assuming PutObject returns success. I am bit hesitant to  agree with that. My recommendation, PutObject should return something from which actual written data can be found along with error. Just returning error and assuming it did it job, is not the correct way.

---

*[pkg/miniogw/gateway-storj.go, line 291 at r1](https://reviewable.io/reviews/storj/storj/117#-LGQlNNPCUWHwCWEGqaL:-LGVusjg0e8miPtKAEy6:b-896fix) ([raw file](https://github.com/storj/storj/blob/78a99445889c7762bb31fb6dc31c6897cc1fef8c/pkg/miniogw/gateway-storj.go#L291)):*
<details><summary><i>Previously, aligeti wrote…</i></summary><blockquote>

resolution agreed to remove the functions that create dummy data. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/storj/storj/117)*
<!-- Sent from Reviewable.io -->

jtolds commented on pkg/miniogw/gateway-storj.go:
```diff
@@ -199,107 +98,67 @@ func (s *storjObjects) GetBucketInfo(ctx context.Context, bucket string) (
+	test := &mpb.StorjMetaInfo{
+		Metadata: metadata,
 	}
+	metainfo, err := proto.Marshal(test)
```
unchecked error

jtolds commented on protos/objects/meta.proto:
```diff
@@ -0,0 +1,41 @@
+	string StorageClass= 9; 
+
+    //add custom fields here 
+	map<string, string> Metadata= 10;
```
nice
